src/requirements.cpp

```diff
@@ -965,39 +966,35 @@ void requirement_data::blacklist_item( const itype_id &id )
 }
 
 template <typename T>
-static void apply_replacement( std::vector<std::vector<T>> &vec, const itype_id &id,
-                               const itype_id &replacement )
-{
-    // If the target and replacement are both present, remove the target.
-    // If only the target is present, replace it.
-    for( auto &opts : vec ) {
-        typename std::vector<T>::iterator target = opts.end();
-        typename std::vector<T>::iterator replacement_target = opts.end();
-        for( typename std::vector<T>::iterator iter = opts.begin(); iter != opts.end(); ++iter ) {
-            if( iter->type == id ) {
-                target = iter;
-            } else if( iter->type == replacement ) {
-                replacement_target = iter;
+static void apply_replacements( std::vector<std::vector<T>> &vec,
+                                const std::unordered_map<itype_id, itype_id> &replacements )
+{
+    for( std::vector<T> &opts : vec ) {
+        for( typename std::vector<T>::iterator iiter = opts.begin(); iiter != opts.end(); ) {
+            auto riter = replacements.find( iiter->type );
+            if( riter != replacements.cend() ) {
+                itype_id to_id = riter->second;
+                // Replace an item or outright remove it if its replacement
+                // is already present in the vector, to prevent duplicates
+                bool has_duplicates = std::count_if( opts.cbegin(), opts.cend(), [&to_id]( const T & item ) {
+                    return item.type == to_id;
+                } ) != 0;
```

Comment:
I would have used `find_if` here, which has the advantage of early exit.  On the other hand, `count_if` is less verbose, so I can see the argument for it.